pandas/_libs/tslibs/conversion.pyx

```diff
@@ -220,6 +220,14 @@ cdef int64_t cast_from_unit(
             f"cannot convert input {ts} with the unit '{unit}'"
         ) from err
 
+    # In the event that ts itself is already an integer type, cast it to a
+    # 64-bit integer to ensure that frac will also be a 64-bit integer after
+    # type promotion changes in NEP 50. If this is not done, for example,
+    # np.int32 values for ts can lead to np.int32 values for frac, which can
+    # raise unexpected overflow errors downstream (GH 56996).
+    if isinstance(ts, np.integer):
```

Comment:
Can we not just declare `frac` to be int64 if that is what is required? Generally we are casting too much in this function. I am also surprised that `frac - base` doesn't follow the C implicit conversion rules, since `base` is int64 at this point; I feel like simplifying this and using explicit types would help across the board

Reply:
`frac` is mainly needed for when `ts` is a floating point value, in which case it also needs to be a floating point value, so we cannot declare it to be a 64-bit integer up front.

In thinking about the floating point case, it occurs to me that NEP 50 also brings slight answer changes in that context as well.  E.g. with NumPy < 2:

```
>>> pd.to_timedelta(np.float32(3.2), "D")
Timedelta('3 days 04:48:00.004147200')
```

and with NumPy >= 2:

```
>>> pd.to_timedelta(np.float32(3.2), "D")
Timedelta('3 days 04:48:00.005046272')
```

Again this is due to changes in type promotion rules for `frac = ts - base`.  Previously `np.float32(3.2) - 3` would return a 64-bit float, but with NumPy >= 2 it returns a 32-bit float.  

Is that something we would also like to address?

Reply:
I think what would help this function is giving `frac` a type, and maybe making another variable if we need  branching. Mixing Python objects into mostly "typed" Cython code like this is tough to follow

Reply:
Thanks and sorry for the delay in getting back.  I pushed an update in 6297494—let me know if that is along the lines of what you were thinking.

Reply:
Ah it seems like declaring `frac_float64` to be a `float64_t` type [changes answers in the 32-bit build](https://github.com/pandas-dev/pandas/actions/runs/8679648945/job/23798733781?pr=57984).  Do you have any suggestions for how to address that?

Reply:
Does it work if you just use the `float` data type? Also can you get rid of the `isinstance` check with the current design?

Reply:
I tried and unfortunately it seems like it does not—I think `float` locks it to always be a float32 value in Cython.  

I have not taken the time to fully understand it yet—I'm admittedly a bit confused—but in my testing in a Docker container the only approach that seems to work on 32-bit platforms is the one where the result of `ts - base` is assigned to a variable with an undeclared type.

Reply:
I am somewhat surprised there is no build warnings to hint us at the issue. @lithomas1 is there something else we need to do to make build warnings visible in CI?